message.go

```diff
@@ -171,6 +173,42 @@ func (m *MessageEdit) SetEmbed(embed *MessageEmbed) *MessageEdit {
 	return m
 }
 
+// AllowedMentionType describes the types of mentions used
+// in the MessageAllowedMentions type.
+type AllowedMentionType string
+
+// The types of mentions used in MessageAllowedMentions.
+const (
+	AllowedMentionTypeRoles    AllowedMentionType = "roles"
+	AllowedMentionTypeUsers    AllowedMentionType = "users"
+	AllowedMentionTypeEveryone AllowedMentionType = "everyone"
+)
+
+// MessageAllowedMentions allows the user to specify which mentions
+// Discord is allowed to parse in this message. This is useful when
+// sending user input as a message, as it prevents unwanted mentions.
+// If this type is used, all mentions must be explicitly whitelisted,
+// either by putting an AllowedMentionType in the Parse slice
+// (allowing all mentions of that type) or, in the case of roles and
+// users, explicitly allowing those mentions on an ID-by-ID basis.
+// For more information on this functionality, see:
+// https://discordapp.com/developers/docs/resources/channel#allowed-mentions-object-allowed-mentions-reference
+type MessageAllowedMentions struct {
```

Comment:
If one were to try to allow mentions in a message for a particular role (likewise for particular users), would this struct look like the following?

```go
messageAllowedMentionsCase1 := &MessageAllowedMentions{
    Roles: []string{"someRoleID"},
}
```

If so, then this comment is a little confusing to me because it seems to imply that no mentions would be allowed:

```go
	// The mention types that are allowed to be parsed in this message.
	// Please note that this is purposely **not** marked as omitempty,
	// so if a zero-value MessageAllowedMentions object is provided no
	// mentions will be allowed.
	Parse []AllowedMentionType `json:"parse"`
```

If instead the struct should look like:

```go
messageAllowedMentionsCase2 := &MessageAllowedMentions{
    Parse: []AllowedMentionType{AllowedMentionTypeRoles},
    Roles: []string{"someRoleID"},
}
```

Then this comment becomes a bit confusing because it states we should not use `AllowedMentionTypeRoles` in the `Parse` slice:

```go
	// A list of role IDs to allow. This cannot be used when specifying
	// AllowedMentionTypeRoles in the Parse slice.
	Roles []string `json:"roles,omitempty"`
```

Based on the Discord docs, this is a bit convoluted but the former case (`messageAllowedMentionsCase1`) appears to be the intention. They also have an example when the `parse` field is not present in the resultant JSON:

```json
{
    "content": "<@123> Time for some memes.",
    "allowed_mentions": {
        "users": ["123", "125"]
    }
}
```

Perhaps we do want to allow `omitempty` for `Parse` because if it is not present, it seems to satisfy the intention of disallowing all mention types except for the roles and users that were specifically whitelisted.

Reply:
The intention behind the original comment was to warn the user about the case of a zero-value **`MessageAllowedMentions`** object, not necessarily a zero-value slice; other fields can still be specified.

I do not believe `Parse` can be left as `omit empty`, as then there is no way to disallow everyone and role mentions without specifying a whitelisted user or role; `"allowed_mentions": {}` doesn't disallow them, while `"allowed_mentions": {"parse": []}` does.

Reply:
Yes, leaving `Parse` to `omitempty` can't really be done. There needs to be a good way of disabling all mentions as per the Discord API. If we want to get this as close as possible to the Discord API we can't do much but just warn users that an empty `allowed_mentions` does something different than an empty `allowed_mentions.parse`.